diff --git a/x.c b/x.c
index 2a3bd38..f534347 100644
--- a/x.c
+++ b/x.c
@@ -81,6 +81,7 @@ typedef XftGlyphFontSpec GlyphFontSpec;
 typedef struct {
 	int tty_width, tty_height; /* tty width and height */
 	int w, h; /* window width and height */
+	int hborderpx, vborderpx;
 	int ch; /* char height */
 	int cw; /* char width  */
 	int mode; /* window state/mode flags */
@@ -331,7 +332,7 @@ tty_send(const Arg *arg)
 int
 evcol(XEvent *e)
 {
-	int x = e->xbutton.x - border_pixels;
+	int x = e->xbutton.x - term_window.hborderpx;
 	LIMIT(x, 0, term_window.tty_width - 1);
 	return x / term_window.cw;
 }
@@ -339,7 +340,7 @@ evcol(XEvent *e)
 int
 evrow(XEvent *e)
 {
-	int y = e->xbutton.y - border_pixels;
+	int y = e->xbutton.y - term_window.vborderpx;
 	LIMIT(y, 0, term_window.tty_height - 1);
 	return y / term_window.ch;
 }
@@ -739,6 +740,9 @@ cresize(int width, int height)
 	col = MAX(1, col);
 	row = MAX(1, row);
 
+	term_window.hborderpx = (term_window.w - col * term_window.cw) / 2;
+	term_window.vborderpx = (term_window.h - row * term_window.ch) / 2;
+
 	term_resize(col, row);
 	x_resize(col, row);
 	tty_resize(term_window.tty_width, term_window.tty_height);
@@ -869,8 +873,8 @@ x_hints(void)
 	sizeh->flags = PSize | PResizeInc | PBaseSize | PMinSize;
 	sizeh->height = term_window.h;
 	sizeh->width = term_window.w;
-	sizeh->height_inc = term_window.ch;
-	sizeh->width_inc = term_window.cw;
+	sizeh->height_inc = 1;
+	sizeh->width_inc = 1;
 	sizeh->base_height = 2 * border_pixels;
 	sizeh->base_width = 2 * border_pixels;
 	sizeh->min_height = term_window.ch + 2 * border_pixels;
@@ -1152,8 +1156,8 @@ x_init(int number_cols, int number_rows)
 	x_load_cols();
 
 	/* adjust fixed window geometry */
-	term_window.w = 2 * border_pixels + number_cols * term_window.cw;
-	term_window.h = 2 * border_pixels + number_rows * term_window.ch;
+	term_window.w = 2 * term_window.hborderpx + 2 * border_pixels + number_cols * term_window.cw;
+	term_window.h = 2 * term_window.vborderpx + 2 * border_pixels + number_rows * term_window.ch;
 	if (x_window.gm & XNegative)
 		x_window.l += DisplayWidth(x_window.dpy, x_window.scr) - term_window.w - 2;
 	if (x_window.gm & YNegative)
@@ -1242,7 +1246,7 @@ x_init(int number_cols, int number_rows)
 int
 x_make_glyph_font_specs(XftGlyphFontSpec *specs, const Glyph *glyphs, int len, int x, int y)
 {
-	float winx = border_pixels + x * term_window.cw, winy = border_pixels + y * term_window.ch, xp, yp;
+	float winx = term_window.hborderpx + x * term_window.cw, winy = term_window.vborderpx + y * term_window.ch, xp, yp;
 	ushort mode, prevmode = USHRT_MAX;
 	Font *font = &draw_context.font;
 	int frcflags = FRC_NORMAL;
@@ -1375,7 +1379,7 @@ void
 x_draw_glyph_font_specs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, int y)
 {
 	int charlen = len * ((base.mode & ATTR_WIDE) ? 2 : 1);
-	int winx = border_pixels + x * term_window.cw, winy = border_pixels + y * term_window.ch,
+	int winx = term_window.hborderpx + x * term_window.cw, winy = term_window.vborderpx + y * term_window.ch,
 	    width = charlen * term_window.cw;
 	Color *fg, *bg, *temp, revfg, revbg, truefg, truebg;
 	XRenderColor colfg, colbg;
@@ -1465,17 +1469,17 @@ x_draw_glyph_font_specs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, i
 
 	/* Intelligent cleaning up of the borders. */
 	if (x == 0) {
-		x_clear(0, (y == 0)? 0 : winy, border_pixels,
+		x_clear(0, (y == 0)? 0 : winy, term_window.hborderpx,
 			winy + term_window.ch +
-			((winy + term_window.ch >= border_pixels + term_window.tty_height)? term_window.h : 0));
+			((winy + term_window.ch >= term_window.vborderpx + term_window.tty_height)? term_window.h : 0));
 	}
-	if (winx + width >= border_pixels + term_window.tty_width) {
+	if (winx + width >= term_window.hborderpx + term_window.tty_width) {
 		x_clear(winx + width, (y == 0)? 0 : winy, term_window.w,
-			((winy + term_window.ch >= border_pixels + term_window.tty_height)? term_window.h : (winy + term_window.ch)));
+			((winy + term_window.ch >= term_window.vborderpx + term_window.tty_height)? term_window.h : (winy + term_window.ch)));
 	}
 	if (y == 0)
-		x_clear(winx, 0, winx + width, border_pixels);
-	if (winy + term_window.ch >= border_pixels + term_window.tty_height)
+		x_clear(winx, 0, winx + width, term_window.vborderpx);
+	if (winy + term_window.ch >= term_window.vborderpx + term_window.tty_height)
 		x_clear(winx, winy + term_window.ch, winx + width, term_window.h);
 
 	/* Clean up the region we want to draw to. */
@@ -1569,35 +1573,35 @@ x_draw_cursor(int cx, int cy, Glyph g, int ox, int oy, Glyph og)
 		case 3: /* Blinking Underline */
 		case 4: /* Steady Underline */
 			XftDrawRect(x_window.draw, &drawcol,
-					border_pixels + cx * term_window.cw,
-					border_pixels + (cy + 1) * term_window.ch - \
+					term_window.hborderpx + cx * term_window.cw,
+					term_window.vborderpx + (cy + 1) * term_window.ch - \
 						cursorthickness,
 					term_window.cw, cursorthickness);
 			break;
 		case 5: /* Blinking bar */
 		case 6: /* Steady bar */
 			XftDrawRect(x_window.draw, &drawcol,
-					border_pixels + cx * term_window.cw,
-					border_pixels + cy * term_window.ch,
+					term_window.hborderpx + cx * term_window.cw,
+					term_window.vborderpx + cy * term_window.ch,
 					cursorthickness, term_window.ch);
 			break;
 		}
 	} else {
 		XftDrawRect(x_window.draw, &drawcol,
-				border_pixels + cx * term_window.cw,
-				border_pixels + cy * term_window.ch,
+				term_window.hborderpx + cx * term_window.cw,
+				term_window.vborderpx + cy * term_window.ch,
 				term_window.cw - 1, 1);
 		XftDrawRect(x_window.draw, &drawcol,
-				border_pixels + cx * term_window.cw,
-				border_pixels + cy * term_window.ch,
+				term_window.hborderpx + cx * term_window.cw,
+				term_window.vborderpx + cy * term_window.ch,
 				1, term_window.ch - 1);
 		XftDrawRect(x_window.draw, &drawcol,
-				border_pixels + (cx + 1) * term_window.cw - 1,
-				border_pixels + cy * term_window.ch,
+				term_window.hborderpx + (cx + 1) * term_window.cw - 1,
+				term_window.vborderpx + cy * term_window.ch,
 				1, term_window.ch - 1);
 		XftDrawRect(x_window.draw, &drawcol,
-				border_pixels + cx * term_window.cw,
-				border_pixels + (cy + 1) * term_window.ch - 1,
+				term_window.hborderpx + cx * term_window.cw,
+				term_window.vborderpx + (cy + 1) * term_window.ch - 1,
 				term_window.cw, 1);
 	}
 }
-- 
2.37.1

