diff --git a/config.def.h b/config.def.h
index 20761ef..b12bf81 100644
--- a/config.def.h
+++ b/config.def.h
@@ -117,34 +117,56 @@ float alpha = 0.8;
 /* Terminal colors (16 first used in escape sequence) */
 static const char *colorname[] = {
 	/* 8 normal colors */
-	"black",
-	"red3",
-	"green3",
-	"yellow3",
-	"blue2",
-	"magenta3",
-	"cyan3",
-	"gray90",
+	"#000000",
+	"#ff0000",
+	"#00ff00",
+	"#ffff00",
+	"#0088ff",
+	"#ff00ff",
+	"#00aaaa",
+	"#ffffff",
 
 	/* 8 bright colors */
-	"gray50",
-	"red",
-	"green",
-	"yellow",
-	"#5c5cff",
-	"magenta",
-	"cyan",
-	"white",
-
+	"#333333",
+	"#ff6600",
+	"#00cc00",
+	"#ffbb00",
+	"#0000ff",
+	"#c600c6",
+	"#0066aa",
+	"#f1f1f1",
+
+	/* 18 transparent colors */
+	"#000000",
+	"#440000",
+	"#004400",
+	"#444400",
+	"#000044",
+	"#440044",
+	"#004444",
+	"#444444",
+
+	"#881000",
+	"#882000",
+	"#883000",
+	"#884000",
+	"#885000",
+	"#886000",
+	"#887000",
+	"#887800",
+	"#888000",
+	"#888800",
+ 
 	[255] = 0,
 
 	/* more colors can be added after 255 to use with DefaultXX */
-	"#cccccc",
+ 	"#000000",
+	"#ffff00",
+	"#0000ff",
 	"#555555",
-	"gray90", /* default foreground colour */
-	"black", /* default background colour */
 };
 
+const int trans_colors = 18;
 
 /*
  * Default colors (colorname index)
diff --git a/x.c b/x.c
index 9ec8bc7..6ffbd5c 100644
--- a/x.c
+++ b/x.c
@@ -796,7 +796,7 @@ x_load_color(int i, const char *name, Color *ncolor)
 	XRenderColor color = { .alpha = 0xffff };
 
 	if (!name) {
-		if (BETWEEN(i, 16, 255)) { /* 256 color */
+		if (BETWEEN(i, 16 + trans_colors, 255)) { /* 256 color */
 			if (i < 6*6*6+16) { /* same colors as xterm */
 				color.red   = sixd_to_16bit( ((i-16)/36)%6 );
 				color.green = sixd_to_16bit( ((i-16)/6) %6 );
@@ -841,6 +841,12 @@ x_load_cols(void)
 	draw_context.col[default_background].color.alpha = (unsigned short)(0xffff * alpha);
 	draw_context.col[default_background].pixel &= 0x00FFFFFF;
 	draw_context.col[default_background].pixel |= (unsigned char)(0xff * alpha) << 24;
+
+	for (i = 16; i < 16 + trans_colors; i++) {
+		draw_context.col[i].color.alpha = (unsigned short)(0xffff * alpha);
+		draw_context.col[i].pixel &= 0x00FFFFFF;
+		draw_context.col[i].pixel |= (unsigned char)(0xff * alpha) << 24;
+	}
 	loaded = 1;
 }
 
